framework/src/relationshipmanagers/GhostLowerDElems.C

```diff
@@ -0,0 +1,78 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+// App includes
+#include "GhostLowerDElems.h"
+#include "Executioner.h"
+#include "FEProblemBase.h"
+#include "MooseApp.h"
+
+// libMesh includes
+#include "libmesh/elem.h"
+#include "libmesh/mesh_base.h"
+#include "libmesh/boundary_info.h"
+
+registerMooseObject("MooseApp", GhostLowerDElems);
+
+using namespace libMesh;
+
+InputParameters
+GhostLowerDElems::validParams()
+{
+  InputParameters params = RelationshipManager::validParams();
+  params.set<bool>("attach_geometric_early") = false;
+  return params;
+}
+
+GhostLowerDElems::GhostLowerDElems(const InputParameters & params) : RelationshipManager(params) {}
+
+GhostLowerDElems::GhostLowerDElems(const GhostLowerDElems & other) : RelationshipManager(other) {}
+
+std::string
+GhostLowerDElems::getInfo() const
+{
+  std::ostringstream oss;
+  oss << "GhostLowerDElems";
+  return oss.str();
+}
+
+void
+GhostLowerDElems::operator()(const MeshBase::const_element_iterator & range_begin,
+                             const MeshBase::const_element_iterator & range_end,
+                             const processor_id_type p,
+                             map_type & coupled_elements)
+{
+  mooseAssert(_moose_mesh,
+              "The MOOSE mesh must be non-null in order for this relationship manager to work.");
+
+  static const CouplingMatrix * const null_mat = nullptr;
+
+  for (const Elem * const elem : as_range(range_begin, range_end))
+    for (const auto s : elem->side_index_range())
+    {
+      const Elem * const neighbor = elem->neighbor_ptr(s);
+      const bool elem_owns_lowerd = !neighbor || elem->id() < neighbor->id();
```

Comment:
There is storage on `MooseMesh` for querying for a lower-dimensional element from a higher-dimensional element and side. However, we only store a map from higher-d element and side to lower-d element for *one* of the higher dimensional elements on an internal side. So we choose to store with a higher-dimensional element key based on which higher-dimensional element that "shares" the lower-dimensional element has a lower ID. So basically if either

1. there is no neighbor -> this higher dimensional element definitely is the key in the moose mesh map
2. there is a neighbor, but it has a higher ID than us -> we still are the key in the moose mesh map. If the neighbor has the lower ID, then *it* is the key in the moose mesh map

Reply:
The logic seems to allow for coupling more than one higher-d neighbor, but not a real case?

Reply:
I don't think I fully understand the question

Reply:
> we only store a map from higher-d element and side to lower-d element for one of the higher dimensional elements on an internal side

How many higher-d elements that have a side defining a lower-d element can we have?

Reply:
It will be proportional to the number of elements in the mesh. These lower-d elements that the Griffin team uses for the hybrid finite-element method exist on every single internal side in the mesh

Reply:
For a replicated mesh, the number of keys in the `_higher_d_elem_side_to_lower_d_elem` map will be equal to the number of faces in the mesh